src/compiler/crystal/syntax/token.cr

```diff
@@ -2,7 +2,180 @@ require "./location"
+      OP_PERIOD_PERIOD_PERIOD     # ...
+      OP_SLASH                    # /
+      OP_SLASH_SLASH              # //
+      OP_SLASH_SLASH_EQ           # //=
+      OP_SLASH_EQ                 # /=
+      OP_COLON                    # :
+      OP_COLON_COLON              # ::
+      OP_SEMICOLON                # ;
+      OP_LT                       # <
+      OP_LT_LT                    # <<
+      OP_LT_LT_EQ                 # <<=
+      OP_LT_EQ                    # <=
+      OP_LT_EQ_GT                 # <=>
+      OP_EQ                       # =
+      OP_EQ_EQ                    # ==
+      OP_EQ_EQ_EQ                 # ===
+      OP_EQ_GT                    # =>
+      OP_EQ_TILDE                 # =~
+      OP_GT                       # >
+      OP_GT_EQ                    # >=
+      OP_GT_GT                    # >>
+      OP_GT_GT_EQ                 # >>=
+      OP_QUESTION                 # ?
+      OP_AT_SQUAREL               # @[
+      OP_SQUAREL                  # [
+      OP_SQUAREL_SQUARER          # []
+      OP_SQUAREL_SQUARER_EQ       # []=
+      OP_SQUAREL_SQUARER_QUESTION # []?
+      OP_SQUARER                  # ]
+      OP_CARET                    # ^
+      OP_CARET_EQ                 # ^=
+      OP_GRAVE                    # `
+      OP_CURLYL                   # {
+      OP_CURLYL_PERCENT           # {%
+      OP_CURLYL_CURLYL            # {{
+      OP_BAR                      # |
+      OP_BAR_EQ                   # |=
+      OP_BAR_BAR                  # ||
+      OP_BAR_BAR_EQ               # ||=
+      OP_CURLYR                   # }
+      OP_TILDE                    # ~
+
+      # non-flag enums are special since the `IO` overload relies on the
+      # `String`-returning overload instead of the other way round
+      def to_s : String
+        {% begin %}
+          {%
+            operator1 = {
+              "BANG" => "!", "DOLLAR" => "$", "PERCENT" => "%", "AMP" => "&", "PARENL" => "(",
+              "PARENR" => ")", "STAR" => "*", "PLUS" => "+", "COMMA" => ",", "MINUS" => "-",
+              "PERIOD" => ".", "SLASH" => "/", "COLON" => ":", "SEMICOLON" => ";", "LT" => "<",
+              "EQ" => "=", "GT" => ">", "QUESTION" => "?", "AT" => "@", "SQUAREL" => "[",
+              "SQUARER" => "]", "CARET" => "^", "GRAVE" => "`", "CURLYL" => "{", "BAR" => "|",
+              "CURLYR" => "}", "TILDE" => "~",
+            }
+          %}
+
+          case value
+          {% for member in @type.constants %}
+          when {{ @type.constant(member) }}
+            {% if member.starts_with?("OP_") %}
+              {% parts = member.split("_") %}
+              {{ parts.map { |ch| operator1[ch] || "" }.join("") }}
+            {% elsif member.starts_with?("MAGIC_") %}
+              {{ "__#{member[6..-1].id}__" }}
+            {% else %}
+              {{ member.stringify }}
+            {% end %}
+          {% end %}
+          else
+            value.to_s
+          end
+        {% end %}
```

Comment:
I don't like this generative implementation. I would prefer to spell everything out.
This is much more complex and harder to understand than a simple list of `when OP_CARET then "^"`. It ties constant names closely to their string representation which reduces flexibility. I'd like more flexibility and to consider some improvements to operator names.
And in the end it only saves a couple of LOC, where the saved LOCs would be super trivial.